chrome/content/zotero/zoteroPane.js

```diff
@@ -1893,26 +1909,74 @@ var ZoteroPane = new function()
 		}
 		this.selectItem(item.id).done();
 	}
+
+
+	/**
+	 * Check whether every selected item can be restored from trash
+	 *
+	 * @return {Boolean}
+	 */
+	this.canRestoreSelectedItems = function () {
+		let collectionTreeRow = this.getCollectionTreeRow();
+		if (!collectionTreeRow.isTrash()) {
+			return false;
+		}
+
+		return this.getSelectedItems().some(item => item.deleted);
+	};
 	
 	
 	/**
 	 * @return {Promise}
 	 */
-	this.restoreSelectedItems = Zotero.Promise.coroutine(function* () {
-		var items = this.getSelectedItems();
-		if (!items) {
+	this.restoreSelectedItems = async function () {
+		let items = this.getSelectedItems();
+		if (!items.length) {
 			return;
 		}
-		
-		yield Zotero.DB.executeTransaction(function* () {
-			for (let i=0; i<items.length; i++) {
-				items[i].deleted = false;
-				yield items[i].save({
-					skipDateModifiedUpdate: true
-				});
+
+		let selectedIDs = new Set(items.map(item => item.id));
+		let isSelected = itemOrID => (itemOrID.id ? selectedIDs.has(itemOrID.id) : selectedIDs.has(itemOrID));
+
+		await Zotero.DB.executeTransaction(async () => {
+			let savePromises = [];
+			for (let row = 0; row < this.itemsView.rowCount; row++) {
+				// Only look at top-level items
+				if (this.itemsView.getLevel(row) !== 0) {
+					continue;
+				}
+
+				let parent = this.itemsView.getRow(row).ref;
+
+				if (isSelected(parent)) {
+					if (parent.deleted) {
+						parent.deleted = false;
+						savePromises.push(parent.save());
+					}
+
+					let children = [...parent.getNotes(true), ...parent.getAttachments(true)];
+					let noneSelected = !children.some(isSelected);
+					for (let child of Zotero.Items.get(children)) {
+						Zotero.debug('Child: ' + child.id, -1)
+						if ((noneSelected || isSelected(child)) && child.deleted) {
+							child.deleted = false;
+							savePromises.push(child.save());
+						}
+					}
+				}
+				else {
+					let children = [...parent.getNotes(true), ...parent.getAttachments(true)];
+					for (let child of Zotero.Items.get(children)) {
+						if (isSelected(child) && child.deleted) {
+							child.deleted = false;
+							savePromises.push(child.save());
+						}
+					}
+				}
 			}
-		}.bind(this));
-	});
+			await Promise.all(savePromises);
```

Comment:
We don't do this for object saves. It might be fine, or it might cause weird race conditions, but we don't design or test for it anywhere, and in any case it probably doesn't speed anything up. Can just use `await item.save()` within the loop, despite the eslint warning (which we should probably remove).

This makes more sense for things like network requests where a single request could take anywhere from 0 to many seconds and we don't want other requests to wait, but that's not an issue here.